### PR TITLE
Rewind index to improve error recovery

### PR DESCRIPTION
**Pull request by stasm**
The error recovery logic in the reference parser and the optimistic runtime parser is able to resume parsing at an offset which precedes the exact index of the error. Consider the following example:

```properties
foo = {
bar = Bar
     ^----- Expected token: "}"
```

The error occurs well into the `bar…` line. At the same time, I would expect `bar` to be a functional message and to parse correctly. Messages preceding any given message should not be allowed to break it.

This wasn't easy to achieve with the `ParserStream` architecture based on an iterator over the source string. I re-wrote it to use a simpler cursor-based approach. As it turned out, this simplified `stream.js` quite much and it also improved the performance :)

```
jsshell 63 (ms, compared to zeroseven):
  mean:   28.86 (-28%)
  stdev:  1.35
  sample: 30

node.js 9.11 (ms, compared to zeroseven):
  mean:   8.51 (-32%)
  stdev:  1.05
  sample: 30

```

### CODE REVIEW

File: fluent-syntax/test/fixtures_structure/unclosed.json [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+          "type": "Annotation",
+          "code": "E0003",
+          "args": [
+            "}"
+          ],
+          "message": "Expected token: \"}\"",
+          "span": {
+            "type": "Span",
+            "start": 16,
+            "end": 16
```
[the remark this answers is not in the document]
Comment (stasm) [reply fluent-syntax/test/fixtures_structure/unclosed.json]: If we rewind the error position, pretty printing the annotations (or highlighting them in IDEs) would always point to the beginning of the line. Which I'll admit is sometimes good, and sometimes bad :) It's the choice between:

```
A) Error position unchanged.

# This makes sense to me.
! E0003 on line 2:
  | foo = {
  | invalid placeable }
  …         ^----- Expected token: "}"

# Meh, this is kind of misleading.
! E0003 on line 2:
  | foo = {
  | bar = Bar
  …     ^----- Expected token: "}"
```

And: 

```
B) Error position rewound.

# Meh, we could do better.
! E0003 on line 2:
  | foo = {
  | invalid placeable }
  … ^----- Expected token: "}"

# This makes sense to me.
! E0003 on line 2:
  | foo = {
  | bar = Bar
  … ^----- Expected token: "}"
```

I take it that you vote for B, as it's likely more common?

File: fluent-syntax/src/stream.js [hunk 2]
```diff
@@ -1,130 +1,34 @@
...
-      return false;
-    }
-
-    const ret = this.peek();
-
-    this.peekIndex -= 1;
-
-    return ret === ch;
+    this.peekOffset++;
+    return this.currentPeek;
```
[the remark this answers is not in the document]
Comment (stasm) [reply fluent-syntax/src/stream.js]: I see what you mean. There's no difference in the perf benchmarks, but I like the idea of making the implementation symmetric. Also, I think I'll ditch the pre-increment operator in `next()`. It's easy to miss when reading the code.

File: fluent-syntax/src/parser.js [hunk 3]
```diff
@@ -134,9 +134,12 @@ export default class FluentParser {
         throw err;
       }
 
-      const errorIndex = ps.index;
+      let errorIndex = ps.index;
       ps.skipToNextEntryStart(entryStartPos);
-      const nextEntryStart = ps.index;
+      let nextEntryStart = ps.index;
```
Comment (Pike) [inline fluent-syntax/src/parser.js]: Nit, this can still be a `const`, right?

Comment (stasm) [reply fluent-syntax/src/parser.js]: Ah, yes. I stopped using `const` in most of the code I write these days and only use them for actual constants. I typed `let` out of habit. I'll revert to `const` for consistency with the rest of the file.

**Pull request merged**